ENGAGEHF/Dashboard/Vitals/VitalsManager.swift

```diff
@@ -0,0 +1,311 @@
+        heartRateHistory.max { $0.startDate < $1.startDate }
+    }
+    public var latestBloodPressure: HKCorrelation? {
+        bloodPressureHistory.max { $0.startDate < $1.startDate }
+    }
+    public var latestWeight: HKQuantitySample? {
+        weightHistory.max { $0.startDate < $1.startDate }
+    }
+    
+    
+    /// Call on initial configuration:
+    /// - Add a snapshot listener to the three health data collections
+    public func configure() {
+        if ProcessInfo.processInfo.isPreviewSimulator {
+            self.setupPreview()
+            return
+        }
+        
+        authStateDidChangeListenerHandle = Auth.auth().addStateDidChangeListener { [weak self] _, user in
+            self?.registerSnapshotListeners(user: user)
+        }
+        
+        self.registerSnapshotListeners(user: Auth.auth().currentUser)
+    }
+    
+    private func registerSnapshotListeners(user: User?) {
+        self.logger.debug("Initializing vitals snapshot listener...")
+        
+        // Remove previous snapshot listeners for the user before creating new ones
+        for prevListener in self.snapshotListeners {
+            prevListener.remove()
+        }
+        self.snapshotListeners = []
+        
+        // Only register snapshot listeners when a user is signed in
+        guard let uid = user?.uid else {
+            self.logger.debug("No user signed in, skipping snapshot listener.")
+            return
+        }
+        
+        let firestore = Firestore.firestore()
+        let userDocRef = firestore.collection("users").document(uid)
+        
+        // Weight snapshot listener
+        self.snapshotListeners.append(
+            self.registerSnapshot(
+                collectionReference: userDocRef.collection("bodyWeightObservations"),
+                storage: \.weightHistory,
+                mapObservation: convertToHKQuantitySample
+            )
+        )
+        
+        // Heart Rate snapshot listener
+        self.snapshotListeners.append(
+            self.registerSnapshot(
+                collectionReference: userDocRef.collection("heartRateObservations"),
+                storage: \.heartRateHistory,
+                mapObservation: convertToHKQuantitySample
+            )
+        )
+        
+        // Blood Pressure snapshot listener
+        self.snapshotListeners.append(
+            self.registerSnapshot(
+                collectionReference: userDocRef.collection("bloodPressureObservations"),
+                storage: \.bloodPressureHistory,
+                mapObservation: convertToHKCorrelation
+            )
+        )
+    }
+    
+    private func registerSnapshot<T>(
+        collectionReference: CollectionReference,
+        storage: ReferenceWritableKeyPath<VitalsManager, [T]>,
+        mapObservation: @escaping (R4Observation) throws -> T
+    ) -> ListenerRegistration {
+        // Listen for Weight measurements
+        let listener = collectionReference
```

Comment:
We could remove the local storage of the listener if we return it right after this statement. We could even omit the return statement.